test/integration/scheduler_perf/perf_util.go

```diff
@@ -0,0 +1,348 @@
+// recent scrape.
+//
+// Monotonicity is usually guaranteed because if a bucket with upper bound
+// u1 has count c1, then any bucket with a higher upper bound u > u1 must
+// have counted all c1 observations and perhaps more, so that c  >= c1.
+//
+// Randomly interspersed partial sampling breaks that guarantee, and rate()
+// exacerbates it. Specifically, suppose bucket le=1000 has a count of 10 from
+// 4 samples but the bucket with le=2000 has a count of 7 from 3 samples. The
+// monotonicity is broken. It is exacerbated by rate() because under normal
+// operation, cumulative counting of buckets will cause the bucket counts to
+// diverge such that small differences from missing samples are not a problem.
+// rate() removes this divergence.)
+//
+// bucketQuantile depends on that monotonicity to do a binary search for the
+// bucket with the φ-quantile count, so breaking the monotonicity
+// guarantee causes bucketQuantile() to return undefined (nonsense) results.
+//
+// As a somewhat hacky solution until ingestion is atomic per scrape, we
+// calculate the "envelope" of the histogram buckets, essentially removing
+// any decreases in the count between successive buckets.
+
+func ensureMonotonic(buckets buckets) {
+	max := buckets[0].count
+	for i := range buckets[1:] {
+		switch {
+		case buckets[i].count > max:
+			max = buckets[i].count
+		case buckets[i].count < max:
+			buckets[i].count = max
+		}
+	}
+}
+
+type prometheusCollector struct {
+	metric string
+	cache  *dto.MetricFamily
+}
+
+func newPrometheusCollector(metric string) *prometheusCollector {
+	return &prometheusCollector{
+		metric: metric,
+	}
+}
+
+func (pc *prometheusCollector) collect() *DataItem {
+	var metricFamily *dto.MetricFamily
+	m, _ := legacyregistry.DefaultGatherer.Gather()
+	for _, mFamily := range m {
+		if mFamily.Name == nil || *mFamily.Name != pc.metric {
+			continue
+		}
+		metricFamily = mFamily
+		break
+	}
+
+	if metricFamily == nil {
+		klog.Infof("Metric %q not found", pc.metric)
+		return nil
+	}
+
+	dataItem := pc.promHist2Summary(metricFamily.GetMetric()[0].GetHistogram())
+	if dataItem.Data == nil {
+		return nil
+	}
+
+	// clear the metrics since it's shared among all tests run inside the same binary
+	clearPromHistogram(metricFamily.GetMetric()[0].GetHistogram())
+
+	return dataItem
+}
+
+func (pc *prometheusCollector) promHist2Summary(hist *dto.Histogram) *DataItem {
+	buckets := []bucket{}
+
+	if hist.SampleCount != nil && *hist.SampleCount == 0 {
+		return &DataItem{}
+	}
+
+	for _, bckt := range hist.Bucket {
+		b := bucket{
+			count:      float64(*bckt.CumulativeCount),
+			upperBound: *bckt.UpperBound,
+		}
+		buckets = append(buckets, b)
+	}
+
+	q50 := bucketQuantile(0.50, buckets)
+	q90 := bucketQuantile(0.90, buckets)
+	q99 := bucketQuantile(0.95, buckets)
+
+	return &DataItem{
+		Labels: map[string]string{
+			"Metric": pc.metric,
+		},
+		Data: map[string]float64{
```

Comment:
We should have an `Average` as well.

Reply:
What would be the average in case of a histogram with an exponential bucket length? Is `Perc50` insufficient?

Reply:
Prometheus gives you `sum` and `count`, see https://prometheus.io/docs/concepts/metric_types/#histogram.

It's useful to get the average latency of the scheduling cycle as a cross check for throughput, i.e., `average latency = 1/throughput`.

I am OK to defer this.

Reply:
Done

Reply:
Example:
```json
{
  "data": {
    "Average": 17.701317640933336,
    "Perc50": 13.656216216216217,
    "Perc90": 31.608961303462323,
    "Perc99": 53.44262295081967
  },
  "unit": "ms",
  "labels": {
    "Metric": "scheduler_e2e_scheduling_duration_seconds",
    "Name": "SchedulingBasic/5000Nodes/5000InitPods/1000PodsToSchedule"
  }
}
```